File: BinaryTree/php/pathFinder.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/TreeNode.php';

function pathFinder(?TreeNode $root, int $target): ?array
{
    $path = buildPath($root, $target);
    if ($path === null) {
        return null;
    }
    return array_reverse($path);
}

function buildPath(?TreeNode $node, int $target): ?array
{
    if ($node === null) {
        return null;
    }

    if ($node->val === $target) {
        return [$node->val];
    }

    $left = buildPath($node->left, $target);
    if ($left !== null) {
        $left[] = $node->val;
        return $left;
    }

    $right = buildPath($node->right, $target);
    if ($right !== null) {
        $right[] = $node->val;
        return $right;
    }

    return null;
}

File: BinaryTree/go/pathFinder.go
package binarytree

// PathFinder returns the path from root to target, if it exists.
func PathFinder(root *TreeNode, target int) ([]int, bool) {
	path := buildPath(root, target)
	if path == nil {
		return nil, false
	}

	for i, j := 0, len(path)-1; i < j; i, j = i+1, j-1 {
		path[i], path[j] = path[j], path[i]
	}
	return path, true
}

func buildPath(node *TreeNode, target int) []int {
	if node == nil {
		return nil
	}
	if node.Val == target {
		return []int{node.Val}
	}

	if left := buildPath(node.Left, target); left != nil {
		return append(left, node.Val)
	}
	if right := buildPath(node.Right, target); right != nil {
		return append(right, node.Val)
	}
	return nil
}
